utils/src/arg_parser.rs

```diff
@@ -0,0 +1,582 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::Borrow;
+use std::collections::HashSet;
+use std::env;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::process;
+use std::result;
+
+pub type Result<T> = result::Result<T, Error>;
+
+const DOUBLE_DASH: &str = "--";
+const HELP_ARG: &str = "--help";
+const FLAG_PROVIDED: &str = "true";
+
+/// Errors associated with parsing and validating arguments.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// The required argument was not provided.
+    MissingArgument(String),
+    /// A value for the argument was not provided.
+    MissingValue(String),
+    /// The provided argument was not expected.
+    UnexpectedArgument(String),
+    /// The argument was provided more than once.
+    DuplicatedArgument(String),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            MissingArgument(ref arg) => write!(f, "Argument '{}' required, but not found.", arg),
+            MissingValue(ref arg) => write!(
+                f,
+                "The argument '{}' requires a value, but none was supplied.",
+                arg
+            ),
+            UnexpectedArgument(ref arg) => write!(
+                f,
+                "Found argument '{}' which wasn't expected, or isn't valid in this context.",
+                arg
+            ),
+            DuplicatedArgument(ref arg) => {
+                write!(f, "The argument '{}' was provided more than once.", arg)
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq)]
+pub struct ArgInfo<'a> {
+    name: &'a str,
+    required: bool,
+    requires: Option<&'a str>,
+    takes_value: bool,
+    default_value: Option<&'a str>,
+    optional: bool,
+    help: Option<&'a str>,
+    user_value: Option<String>,
+}
+
+impl<'a> ArgInfo<'a> {
+    pub fn new(name: &'a str) -> ArgInfo<'a> {
+        ArgInfo {
+            name,
+            required: false,
+            requires: None,
+            takes_value: false,
+            default_value: None,
+            optional: false,
+            help: None,
+            user_value: None,
+        }
+    }
+
+    pub fn required(mut self, required: bool) -> ArgInfo<'a> {
```

Comment:
nit: the explicit lifetime is not needed here (nor for the other setters). You can just do:

```rust
pub fn required(mut self, required: bool) -> Self { ... }
```